core/contracts/Registry.sol

```diff
@@ -57,58 +61,103 @@ contract Registry is RegistryInterface, MultiRole {
-        // of storage reads. This is because storage reads are far more expensive than extra memory space (~100:1).
-        address[] memory tmpDerivativeArray = new address[](registeredDerivatives.length);
-        uint outputIndex = 0;
-        for (uint i = 0; i < registeredDerivatives.length; i = i.add(1)) {
-            address derivative = registeredDerivatives[i];
-            if (derivativesToParties[derivative].parties[party]) {
-                // Copy selected derivative to the temporary array.
-                tmpDerivativeArray[outputIndex] = derivative;
-                outputIndex = outputIndex.add(1);
-            }
-        }
+    function getRegisteredDerivatives(address party) external view returns (address[] memory) {
+        return partyMembersToDerivatives[party].derivatives;
+    }
+
+    function addPartyToDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Can only add to valid derivative"
+        );
+        require(!isPartyMemberOfDerivative(party, derivativeAddress), "Can only register a party once");
+
+        // Push the derivative and store the index.
+        uint derivativeIndex = partyMembersToDerivatives[party].derivatives.push(derivativeAddress);
+        partyMembersToDerivatives[party].derivativeIndex[derivativeAddress] = derivativeIndex - 1;
 
-        // Copy the temp array to the return array that is set to the correct size.
-        derivatives = new address[](outputIndex);
-        for (uint j = 0; j < outputIndex; j = j.add(1)) {
-            derivatives[j] = tmpDerivativeArray[j];
+        emit PartyMemberAdded(derivativeAddress, party);
+    }
+
+    function removePartyFromDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        PartyMember storage partyMember = partyMembersToDerivatives[party];
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Remove only from valid derivative"
+        );
+        require(isPartyMemberOfDerivative(party, derivativeAddress), "Can only remove an existing party member");
+
+        // Index of the current location of the derivative to remove.
+        uint deleteIndex = partyMember.derivativeIndex[derivativeAddress];
+
+        // Set the removed derivative address in the lookup map to 0 (deleted).
+        partyMember.derivativeIndex[derivativeAddress] = 0;
+
+        uint256 numberOfDerivatives = partyMember.derivatives.length;
+        require(numberOfDerivatives != 0, "Can't remove if party has no derivatives");
+
+        // If there is more than one derivative then swap the derivative to be removed with the
+        // last position in the array and delete the last position.
+        if (numberOfDerivatives > 1) {
```

Comment:
Curious to hear your thoughts on an alternative strategy here. What if we were to try to eliminate this `if` entirely, so there are no special cases? Since this is fairly involved/subtle state-changing logic, I, personally, have a slight preference toward making the logic as streamlined and minimal as possible. The other slight advantage is that there seems to be a little repeated logic in the `if` and `else` blocks that we might be able to eliminate by combining them.

I was roughly thinking that the following could work:
- Move the last element to your deletion slot and overwrite its stored index. (swap step)
- Delete the data at the last element and remove the stored index for the derivative address that you're deleting.
- Decrement the length of the array.

If there's only one element in the array, the swap will cost a few extra operations, but this will impose a fairly small amount of additional gas cost due to net sstore metering added in https://eips.ethereum.org/EIPS/eip-1283.

Once again, just want to hear your thoughts -- we definitely don't have to do it this way.